File: packages/sdk-platforms/rust/zksync-sso/crates/sdk/src/utils/session/session_lib/session_spec/constraint.rs
use crate::{
    contracts::SessionLib::Constraint as SessionLibConstraint,
    utils::session::session_lib::session_spec::{
        condition::Condition, usage_limit::UsageLimit,
    },
};
use alloy::primitives::FixedBytes;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct Constraint {
    pub condition: Condition,
    pub index: u64,
    pub ref_value: FixedBytes<32>,
    pub limit: UsageLimit,
}

impl From<SessionLibConstraint> for Constraint {
    fn from(value: SessionLibConstraint) -> Self {
        Constraint {
            condition: value.condition.try_into().unwrap(),
            index: value.index,
            ref_value: value.refValue,
            limit: value.limit.into(),
        }
    }
}

impl From<Constraint> for SessionLibConstraint {
    fn from(value: Constraint) -> Self {
        SessionLibConstraint {
            condition: value.condition.into(),
            index: value.index,
            refValue: value.ref_value,
            limit: value.limit.into(),
        }
    }
}

File: packages/sdk-platforms/swift/ZKsyncSSO/Sources/ZKsyncSSO/Account/Sessions/SessionLib/SessionSpec/Constraint.swift
import Foundation

public struct Constraint: Codable, Equatable, Hashable, Sendable {
    public var condition: Condition
    public var index: UInt64
    public var refValue: String
    public var limit: UsageLimit

    public init(condition: Condition, index: UInt64, refValue: String, limit: UsageLimit) {
        self.condition = condition
        self.index = index
        self.refValue = refValue
        self.limit = limit
    }
}
